Tmedia 402 logged out nav cta

JackHowa opened the pull request:
## Description
As a user, I can click “Sign Up” or “Log In” CTAs from the navigation.

## Jira Ticket
- [TMEDIA-402](https://arcpublishing.atlassian.net/secure/RapidBoard.jspa?rapidView=875&modal=detail&selectedIssue=TMEDIA-402&assignee=5e387d6a1b1d910e5dfd7a9b)

## Acceptance Criteria
- [x] Sign up URL is a Custom field with a default of account/signup
- [x] Log In URL is a Custom field with a default of account/login
- [x] Clicking Sign Up nav CTA on Desktop takes you to sign up URL
- [x] Clicking Log In nav CTA on Desktop takes you to log in URL
- [x] Single Button on Mobile w/dropdown on click
- [x] Primary font is used -> see reusable button component
- [x] Language keys are used for displaying words -> "Sign Up" and "Log In" are in the intl files
- [x] Accessibility is to be taken into account
E.g. Button when a button, link when a user is being taken to a page -> using a link tag for the sign up links and the "as" to modify the styled component

## Test Steps
- Add test steps a reviewer must complete to test this PR

1. Checkout this branch `git checkout TMEDIA-402-logged-out-nav-cta`
2. Run fusion repo with linked blocks `npx fusion start -f -l @wpmedia/identity-block`
3. Create page with header nav chain
4. Add header nav action to desktop and mobile custom placements 
5. Change screen and see both rendered as expected on published page with "Sign Up" and "Log In" on desktop and an account button with an on-click dropdown on mobile 
6. With custom fields you should be able to change the urls to the page

https://zpl.io/2E85xAn - zeplin docs 

## Effect Of Changes
### Before
_Example: When I clicked the search button, the button turned red._

[include screenshot or gif or link to video, storybook would be awesome]

### After
_Example: When I clicked the search button, the button turned green._

[include screenshot or gif or link to video, storybook would be awesome]

## Dependencies or Side Effects
_Examples of dependencies or side effects are:_
- Additional settings that will be required in the blocks.json
- Changes to the custom fields which will require users to reconfigure features
- Update to css framework or SDK
- Dependency on another PR that needs to be merged first

## Author Checklist
_The author of the PR should fill out the following sections to ensure this PR is ready for review._
- [ ] Confirmed all the test steps a reviewer will follow above are working.
- [ ] Confirmed there are no linter errors. Please run `npm run lint` to check for errors. Often, `npm run lint:fix` will fix those errors and warnings.
- [ ] Ran this code locally and checked that there are not any unintended side effects. For example, that a CSS selector is scoped only to a particular block.
- [ ] Confirmed this PR has reasonable code coverage. You can run `npm run test:coverage` to see your progress.
  - [ ] Confirmed this PR has unit test files
  - [ ] Ran `npm run test`, made sure all tests are passing
  - [ ] If the amount of work to write unit tests for this change are excessive,
please explain why (so that we can fix it whenever it gets refactored).
- [ ] Confirmed relevant documentation has been updated/added.

## Reviewer Checklist
_The reviewer of the PR should copy-paste this template into the review comments on review._

- [ ] Linting code actions have passed.
- [ ] Ran the code locally based on the test instructions.
  - [ ] I don’t think this is needed to be tested locally. For example, a padding style change (storybook?) or a logic change (write a test).
- [ ] I am a member of the engine theme team so that I can approve and merge this. If you're not on the team, you won't have access to approve and merge this pr.
- [ ] Looked to see that the new or changed code has code coverage, specifically. We want the global code coverage to keep on going up with targeted testing.

JackHowa commented on blocks/identity-block/components/Button/index.jsx:
```diff
@@ -0,0 +1,131 @@
+      default:
+        return `
+          background-color: ${getThemeStyle(arcSite)['primary-color']};
+          border-color: ${getThemeStyle(arcSite)['primary-color']};
```
think you're drunk stylelint

nschubach commented on blocks/identity-block/components/Button/styles.scss:
```diff
@@ -0,0 +1,28 @@
+  // border color managed in styled component
+  border-style: solid;
+  border-width: calculateRem(1px);
+
```
Do we need some padding in here for the icon?
![image](https://user-images.githubusercontent.com/2287238/130866789-e890f1ac-9842-4f09-bbc1-a142d50707b8.png)

JackHowa replied:
yep good call -- that's one of my todo items

JackHowa replied:
updated

nschubach commented on blocks/identity-block/components/Button/index.jsx:
```diff
@@ -0,0 +1,131 @@
+      arcSite={arcSite}
+      matchedButtonStyle={matchedButtonStyle}
+      className={`xpmedia-subs-button ${matchedButtonSizeClass}`}
+    >
```
Can we specify a type so it doesn't default to submit?

JackHowa replied:
yep will do -- one of my todos

JackHowa replied:
updated -- thanks @nschubach

matthewroach commented on blocks/identity-block/components/Button/index.jsx:
```diff
@@ -0,0 +1,134 @@
+    buttonType,
+    children,
+    iconType = '',
+    text = '',
```
Text should be a required property.

JackHowa replied:
will do

JackHowa replied:
removed default ''

matthewroach commented on blocks/identity-block/components/Button/index.jsx:
```diff
@@ -0,0 +1,134 @@
+  if (iconType) {
+    switch (iconType) {
+      case 'user':
+        Icon = (
```
We should think about allowing any of our Icon's to be used here, without the need to manually specify them as done here

JackHowa replied:
this is a good idea; I'll make a ticket for it

JackHowa replied:
https://arcpublishing.atlassian.net/browse/TMEDIA-467

matthewroach commented on blocks/identity-block/components/Button/index.jsx:
```diff
@@ -0,0 +1,134 @@
+      className={`xpmedia-subs-button ${matchedButtonSizeClass}`}
+    >
+      {
+        matchedButtonType === BUTTON_TYPES.ICON_ONLY ? (
```
Icon only buttons are valid, but to be accessible will need either the icon to provide the label or for the attribute `aria-label` to be set on the `Button`

JackHowa replied:
ok I added the aria label prop -- thanks. 

should I also change the "context" prop for the icons. I noticed a few months ago we hid the default title and description for the iconshttps://github.com/WPMedia/engine-theme-sdk/blob/stable/src/components/icons/UserIcon.tsx#L10

matthewroach commented on blocks/identity-block/components/Button/index.jsx:
```diff
@@ -0,0 +1,134 @@
+    <StyledDynamicButton
+      arcSite={arcSite}
+      matchedButtonStyle={matchedButtonStyle}
+      className={`xpmedia-subs-button ${matchedButtonSizeClass}`}
```
What makes this button a subscriptions button? `xpmedia-subs-button`? 
While a lot of the reusable components during the subscriptions work we are creating inside of the Identity block - we should think longer term and look at the possibility of this becoming a more widely used component

JackHowa replied:
ok I will rename to `xpmedia-button`

JackHowa replied:
updated

matthewroach commented on blocks/identity-block/components/Button/index.jsx:
```diff
@@ -0,0 +1,134 @@
+  }
+
+  return (
+    <StyledDynamicButton
```
We should allow for custom attributes to be set on the `<button>`

JackHowa replied:
Updated like this? 

```jsx
const StyledDynamicButton = styled.button.attrs((props) => ({
  arcSite: props.arcSite,
  matchedButtonStyle: props.matchedButtonStyle,
}))`
  font-family: ${({ arcSite }) => getThemeStyle(arcSite)['primary-font-family']};
//...
```

matthewroach commented on blocks/identity-block/components/Button/index.story.jsx:
```diff
@@ -0,0 +1,63 @@
+  />
+);
+
+export const UserIconWithLabel = () => (
```
An Icon only story would be helpful

JackHowa replied:
updated

matthewroach commented on blocks/identity-block/components/Button/index.story.jsx:
```diff
@@ -0,0 +1,63 @@
+};
+
+export const Default = () => (
+  <Button text="" />
```
This is not valid, and we shouldn't allow the button to be used in this way

JackHowa replied:
updated -- thank you

Commit:
feat: Add links and link to custom fields for create and login buttons

Commit:
feat: Add translations for logged out header action

Commit:
feat: Use correct margin between buttons

Commit:
feat: Use margin between only links

Commit:
fix: Remove width to fix flex spacing growth

Commit:
feat: Update custom field pages and links

Commit:
fix: Use font-size 14px based on design review

Commit:
feat: Show and hide a tag and button

Commit:
feat: Add dropdown list link items on toggle of mobile button

Commit:
feat: Add overlap on click

Commit:
style: Reorder imports

Commit:
Merge branch 'canary' into TMEDIA-402-logged-out-nav-cta

Commit:
fix: Ensure xpmedia button uses expected width

JackHowa commented on blocks/identity-block/components/HeadlinedSubmitForm/styles.scss:
```diff
@@ -24,6 +24,7 @@
 
   .xpmedia-button {
     margin: map-get($spacers, 'md') 0;
+    width: 100%;
```
updated to match form mocks based on my changes

matthewroach replied:
We shouldn't set the width here - the button should control this - if we add a prop `fullWidth` or similar the button should apply the 100% width - there is likely many other places we will need to set this if it's not part of the button component

JackHowa replied:
updated

nschubach commented on blocks/identity-block/features/header-account-action/default.jsx:
```diff
@@ -64,8 +74,57 @@ const HeaderAccountAction = ({ customFields }) => {
+          onClick={() => setAccountMenu(!isAccountMenuOpen)}
+          // for button accessibility
+          type="button"
+        />
```
`text` and `aria-label` marked required, but none is provided.
![image](https://user-images.githubusercontent.com/2287238/131402286-3bb7b136-a89c-4bea-b014-0f6eb02bc8ff.png)

JackHowa replied:
updated with translations -- thanks

nschubach commented on blocks/identity-block/features/header-account-action/default.jsx:
```diff
@@ -64,8 +74,57 @@ const HeaderAccountAction = ({ customFields }) => {
+            <DropDownLinkListItem
+              href={loginURL}
+              text={phrases.t('identity-block.log-in')}
+              isLastItem
```
Instead of this flag, why aren't we just styling with `&:last-of-type {}` ?

JackHowa replied:
updated -- thanks

nschubach reviewed (dismissed):
Missing some tests:
```
---------------------------------|----------|----------|----------|----------|-------------------|
File                             |  % Stmts | % Branch |  % Funcs |  % Lines | Uncovered Line #s |
---------------------------------|----------|----------|----------|----------|-------------------|
All files                        |    54.55 |    31.25 |    26.67 |       60 |                   |
 header-account-action           |    63.89 |    35.71 |       40 |    71.88 |                   |
  default.jsx                    |    63.89 |    35.71 |       40 |    71.88 |... 4,54,58,67,109 |
 header-account-action/_children |     12.5 |        0 |        0 |     12.5 |                   |
  DropDownLinkListItem.jsx       |    16.67 |        0 |        0 |    16.67 |     9,10,13,18,20 |
  DropdownLinkList.jsx           |        0 |      100 |        0 |        0 |               7,8 |
---------------------------------|----------|----------|----------|----------|-------------------|
```

Looks and works great though!

matthewroach commented on blocks/identity-block/components/Button/index.jsx:
```diff
@@ -148,15 +194,19 @@ Button.propTypes = {
   iconType: PropTypes.oneOf(['user']),
   text: PropTypes.string.isRequired,
   ariaLabel: PropTypes.string.isRequired,
```
`ariaLabel` should not be a required field - The first rule of ARIA is to not use aria. The use of aria-label for this button is when the button is an icon only button - and as `text` is a required prop you could infer the aria-label value from that

JackHowa replied:
ok thanks. I think I understand what you're looking for. updated

JackHowa replied:
remove required 

https://github.com/WPMedia/fusion-news-theme-blocks/pull/1067/files#diff-aeb43ad51c9f6ff9db84b482383910852f8b17f1f16db2ba3e9e109d3106945fR196

fallback to text for aria-label

https://github.com/WPMedia/fusion-news-theme-blocks/pull/1067/files#diff-aeb43ad51c9f6ff9db84b482383910852f8b17f1f16db2ba3e9e109d3106945fR175

matthewroach replied:
We only want to set it if it's really required - and in our use case we only need `aria-label` set when it's an icon button

JackHowa replied:
ok that makes sense

JackHowa replied:
updated -- `aria-label={buttonStyle === BUTTON_TYPES.ICON_ONLY ? (ariaLabel || text) : null}`

sorry for the extra qs

matthewroach commented on blocks/identity-block/components/Button/styles.scss:
```diff
@@ -1,10 +1,14 @@
+  text-decoration: none;
+
+  // override nav header
+  border-radius: calculateRem(4px) !important;
```
Why the need for `!important`?

JackHowa replied:
to override nav header class's border radius

![Screen Shot 2021-08-24 at 16 26 17](https://user-images.githubusercontent.com/5950956/131519546-266f3c1f-f8b4-4769-b8e9-f8bdae70068b.png)

<img width="285" alt="Screen Shot 2021-08-31 at 09 20 15" src="https://user-images.githubusercontent.com/5950956/131519667-79a70bfd-5e53-4a89-8e93-be6417c73664.png">

matthewroach replied:
This is related to the ticket I created - https://arcpublishing.atlassian.net/browse/TMEDIA-468

matthewroach commented on blocks/identity-block/features/header-account-action/_children/DropDownLinkListItem.jsx:
```diff
@@ -0,0 +1,39 @@
+    <li
+      className={`${
+        isLastItem
+          ? 'xpmedia-subs-header-dropdown-list-item--last-item'
```
This could be achieved with CSS

JackHowa replied:
ok will do

matthewroach commented on blocks/identity-block/features/header-account-action/default.jsx:
```diff
@@ -64,8 +74,57 @@ const HeaderAccountAction = ({ customFields }) => {
+          buttonType={BUTTON_TYPES.ICON_ONLY}
+          iconType="user"
+          onClick={() => setAccountMenu(!isAccountMenuOpen)}
+          // for button accessibility
```
This button is controlling a menu and for accessibility will need to have the aria attribute `aria-expanded` with value true or false.
This relates to a previous comment around allowing the button to have additional prop's passed down - we should allow the consumer to pass these attributes down rather than trying to account for all scenarios within our button component

JackHowa replied:
updated

gotcha, my bad, I misunderstood the previous comment. I thought you were talking about the styled component: 

```jsx
const StyledDynamicButton = styled.button.attrs((props) => ({
  arcSite: props.arcSite,
  buttonStyle: props.buttonStyle,
}))`
```

JackHowa replied:
https://github.com/WPMedia/fusion-news-theme-blocks/pull/1067/files#diff-aeb43ad51c9f6ff9db84b482383910852f8b17f1f16db2ba3e9e109d3106945fR174

matthewroach reviewed (changes requested):
Based on the designs all the buttons in the header are using the same styling. It appears as thought the current buttons (Search / Sections Menu) have not been implemented in the past as per design specs.
I've created a ticket to address this - https://arcpublishing.atlassian.net/browse/TMEDIA-468

<img width="791" alt="Microsoft Edge-2021-08-31-09-56 16" src="https://user-images.githubusercontent.com/868127/131473624-1aac212b-2857-4c25-9b9e-c6fbe4d7074f.png">

matthewroach commented on blocks/identity-block/features/header-account-action/default.jsx:
```diff
@@ -64,8 +74,57 @@ const HeaderAccountAction = ({ customFields }) => {
-      <a href={loginURL} className="xpmedia-subs-header--button">Sign In</a>
+    <div className="xpmedia-subs-header--logged-out-header">
+      <div className="xpmedia-subs-header--desktop-logged-out-header">
+        <Button
```
Each button should only render if they have a value set for the `href`

JackHowa replied:
updated. as well for the dropdown links. thanks. will write tests as well

Commit:
fix: Remove required from aria label and add fallback aria label as text

matthewroach commented on blocks/identity-block/components/Button/index.jsx:
```diff
@@ -40,19 +42,41 @@ const StyledDynamicButton = styled.button.attrs((props) => ({
       default:
         return `
           background-color: ${getThemeStyle(arcSite)['primary-color']};
           border-color: ${getThemeStyle(arcSite)['primary-color']};
```
Why not pass in the primary color to avoid all the look ups?

nschubach replied:
If only we had a styled component theme context :p

JackHowa replied:
updated. good refactor. also passed in font-family

matthewroach commented on blocks/identity-block/components/Button/index.jsx:
```diff
@@ -40,19 +42,41 @@ const StyledDynamicButton = styled.button.attrs((props) => ({
           color: ${getThemeStyle(arcSite)['primary-color']};
+
+          &:hover {
+            color: ${getThemeStyle(arcSite)['primary-color']};
```
You could use `color: inherit` here

JackHowa replied:
yeah I was hoping the scoping would be easier as well. seems to be inheriting from another styled component's a:hover though: 

with color: inherit under hover

<img width="110" alt="Screen Shot 2021-08-31 at 10 00 28" src="https://user-images.githubusercontent.com/5950956/131527469-92f69559-a031-448b-855c-644b959989fa.png">

<img width="946" alt="Screen Shot 2021-08-31 at 10 07 29" src="https://user-images.githubusercontent.com/5950956/131527875-7bbd855b-647b-4bf2-835c-a7b104110283.png">


without inherit: 

![Screen Shot 2021-08-31 at 10 05 18](https://user-images.githubusercontent.com/5950956/131527531-e1d5e443-7fa9-41f9-9f4c-1673d696d994.png)
<img width="113" alt="Screen Shot 2021-08-31 at 10 03 29" src="https://user-images.githubusercontent.com/5950956/131527551-c57ab7f3-8749-4585-b61b-78faadce3d0e.png">

Commit:
feat: Pass down any props into the component

Via https://github.com/WPMedia/fusion-news-theme-blocks/pull/1067#discussion_r699123084

Commit:
feat: Only show aria-label when necessary

Via https://github.com/WPMedia/fusion-news-theme-blocks/pull/1067#discussion_r699379553

Commit:
style: Pass in font family and primary color based on feedback

via https://github.com/WPMedia/fusion-news-theme-blocks/pull/1067#discussion_r699377694

Commit:
fix: Only show links if link hrefs are truthy

Via https://github.com/WPMedia/fusion-news-theme-blocks/pull/1067#discussion_r699132301

nschubach commented:
Are we strict on the newline at EOF?

Commit:
feat: Add text and aria label to hide and show identity

Commit:
fix: Refactor to use last of type rather than last classname

Commit:
style: Minor stylelint newline

Commit:
Merge branch 'canary' into TMEDIA-402-logged-out-nav-cta

matthewroach commented on blocks/identity-block/features/header-account-action/_children/DropDownLinkListItem.jsx:
```diff
@@ -0,0 +1,35 @@
+  color: ${(arcSite) => getThemeStyle(arcSite)['primary-color']};
+
+  &:hover {
+    color: ${(arcSite) => getThemeStyle(arcSite)['primary-color']};
```
Should we be consistent and pass in the values for primary-color and font-family?

JackHowa replied:
yep will do

JackHowa replied:
updated

matthewroach commented on blocks/identity-block/components/Button/styles.scss:
```diff
@@ -24,7 +28,7 @@
 }
```
icon container requires `display: flex` to be set to align the icon and text correctly

JackHowa replied:
updated I think in the right place. seems to be already using flex? 

![Screen Shot 2021-09-02 at 10 26 29](https://user-images.githubusercontent.com/5950956/131872556-e7f70bd9-2eb8-4eec-bf37-2bfd1f2470af.png)

matthewroach reviewed (changes requested):
With upcoming work that also has a requirement to use this `button` component we should look to move it to shared styles

Commit:
style: Refactor to pass in colors rather than get theme style

Via https://github.com/WPMedia/fusion-news-theme-blocks/pull/1067#discussion_r700882206

Commit:
Merge branch 'TMEDIA-402-logged-out-nav-cta' of github.com:WPMedia/fusion-news-theme-blocks into TMEDIA-402-logged-out-nav-cta

Commit:
style: Refactor placement of files based on feedback

via https://github.com/WPMedia/fusion-news-theme-blocks/pull/1067#pullrequestreview-744796952

Commit:
Merge branch 'canary' into TMEDIA-402-logged-out-nav-cta

JackHowa commented:
> With upcoming work that also has a requirement to use this `button` component we should look to move it to shared styles

Updated

Commit:
style: Add display flex to icon container per feedback

Update to add flex display

Commit:
Merge branch 'TMEDIA-402-logged-out-nav-cta' of github.com:WPMedia/fusion-news-theme-blocks into TMEDIA-402-logged-out-nav-cta

Commit:
feat: Add fullWidth option for reusability

Commit:
style: Update styled components even though nothing changed

Commit:
test: Add testing and clarify button types and icon types

Commit:
test: Update and test refactor for dropdown link list item

Commit:
style: Remove unused import

JackHowa commented:
updated based on feedback and added tests @matthewroach @nschubach

Commit:
style: Update linting

matthewroach committed:
Update story title

matthewroach committed:
Fix label to be as per state of menu open

matthewroach committed:
Fix button aria-label output

matthewroach committed:
Add intl.json to files array to ensure intl.json file is published

matthewroach committed:
Update label for mobile login options button

No need to provide state in the label as for assistive technologies this is implied by aria-expanded

matthewroach commented:
@JackHowa I made some small changes around the aria-label and accessibility items I noticed. Most notably around how the label is handled for mobile display - as we are using the attribute `aria-expanded` the state of the options (open/closed) is inferred from the attribute. In a screen reader it would notify the user that it's "collapsed" if the value is false, and "expanded"  when open.

matthewroach committed:
Fix font-family being output on button element

matthewroach committed:
default buttons to have type='button'

matthewroach committed:
Remove un needed aria-label

JackHowa commented:
I haven't added the brand-new options one yet, no. But I will do 

Thanks @matthewroach

Commit:
feat: Add intl for all options for login options and login and sign up

Commit:
style: Fix json formatting

JackHowa commented:
@matthewroach updated the translations

matthewroach reviewed (approved):
Nice work @JackHowa
